Add deep linking support to IntentAndroid

satya164 opened the pull request:
Add a method to handle URLs registered to the app,

```js
IntentAndroid.getInitialURL(url => {
    if (url) {
        // do stuff
    }
});
```

Refer - http://developer.android.com/training/app-indexing/deep-linking.html#adding-filters

The API cannot be same as the iOS API (i.e. as a constant), as the activity is not availble at the time of module initialization. Moreover, multiple activties can share the same bridge instance, and the activity itself is not a constant. Hence the initialURL can change.

baoti commented on ReactAndroid/src/main/java/com/facebook/react/modules/intent/IntentModule.java:
```diff
@@ -33,6 +42,26 @@ public String getName() {
+  public Map<String, Object> getConstants() {
+    final Map<String, Object> constants = new HashMap<>();
+
+    Intent intent = ((Activity) mActivityContext).getIntent();
```
Why not mActivity.getIntent() ?

satya164 replied:
There's no `mActivity` here! O.o

baoti replied:
Just change all (activity's) Context to Activity class

baoti replied:
Because it should be a Activity instance.

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/shell/MainReactPackage.java:
```diff
@@ -45,12 +47,18 @@
+
+  public MainReactPackage(Context activityContext) {
+    mActivityContext = activityContext;
+  }
```
I have a diff out to open source the Dialog module which does something similar. Should land it soon-ish.

mkonicek commented on local-cli/generator-android/templates/package/MainActivity.java:
```diff
@@ -25,7 +25,7 @@ protected void onCreate(Bundle savedInstanceState) {
                 .setBundleAssetName("index.android.bundle")
                 .setJSMainModuleName("index.android")
-                .addPackage(new MainReactPackage())
+                .addPackage(new MainReactPackage(this))
```
We need to keep the current constructor around too for backwards-compatibility, otherwise apps won't compile after upgrading.

Have an idea how to get rid of this issue by introducing a base Activity for React apps. Stay tuned :)

mkonicek commented on Libraries/Components/Intent/IntentAndroid.android.js:
```diff
@@ -75,6 +77,16 @@ class IntentAndroid {
 
+  /**
+   * If the app launch was triggered by an app link, it will pop the link url,
+   * otherwise it will return `null`
```
Can you add a link to Android docs that explain how this works?

Thanks for making the API consistent with `LinkingIOS`!

mkonicek commented:
Nice! Thanks for working on this!

satya164 commented:
@mkonicek Thanks. Waiting for the Dialog commit before I change anything here :D

satya164 commented:
@mkonicek I've updated the PR using `getCurrentActivity()` implemented in c06efc0831b9156ee6261d7ddd2086ed87cbc0aa.

However, the current activity is only set on `onResume`, so it's not available while calling `getConstants` method, to have a consistent API between iOS and Android we need that. I've added a method to set the current activity and added the code to the generator to set the activity in `onCreate`. It's backward compatible, but if someone wants the deep linking feature, and has created the project with the old template, he'll need to manually add the code required.

I'm not sure if it's the best approach though. Please have a look and lemme know.

cc @foghina

mkonicek commented:
Just pinged @foghina, thanks for the patience. Many PRs in the queue and branch cut..

mkonicek commented on ReactAndroid/src/main/java/com/facebook/react/modules/intent/IntentModule.java:
```diff
@@ -33,6 +38,27 @@ public String getName() {
+
+    constants.put("initialURL", initialURL);
+
+    return constants;
```
This seems reasonable - the process could be started by an intent.

mkonicek commented on local-cli/generator-android/templates/package/MainActivity.java:
```diff
@@ -22,6 +22,7 @@ protected void onCreate(Bundle savedInstanceState) {
         mReactRootView = new ReactRootView(this);
 
         mReactInstanceManager = ReactInstanceManager.builder()
+                .setActivity(this)
```
@foghina What do you think? You're more familiar with how this works.

foghina commented:
I don't like exposing `setActivity` on `ReactInstanceManager.Builder`. This sets the expectation that the activity will never change, but in fact, it can change. This is why it's called `getCurrentActivity()` and not `getActivity()`. Also, exposing the intent URL as a constant furthers this assumption and will break when you change activities without recreating the `ReactContext`.

Since the JS API is only similar but not identical and since the JS API is a function call anyway (`IntentAndroid.popInitialURL()`), you can easily rewrite it to call into a native method, instead of reading a native constant. Then all the changes in `ReactInstanceManager` and co. become unnecessary.

Exposing constants has bitten us in the past (see screen size), they should only be used for things that are really constant.

satya164 commented:
@foghina @mkonicek Should I change this to a function `getInitialURL` which takes a callback then?

Note that this then diverges from the iOS API. Any thoughts on what should we do in the cross platform module?

satya164 commented:
@foghina @mkonicek Changed to use a callback instead of a constant to get initial URl.

foghina commented on ReactAndroid/src/main/java/com/facebook/react/modules/intent/IntentModule.java:
```diff
@@ -34,24 +34,57 @@ public String getName() {
+
+      callback.invoke(initialURL);
+    } catch (Exception e) {
+      throw new JSApplicationIllegalArgumentException(
```
Can you add an error callback (all `@ReactMethod`s have two callbacks) and call it here instead of throwing an exception? You should probably still log the full exception details, though, to make it easier to debug problems.

satya164 replied:
@foghina Was wondering the same thing. I kept it consistent with the `canOpenURL` method which is in the same file. It's really confusing.

foghina replied:
Oh, darn. Then we can probably leave it as it is for the purposes of this PR but we should refactor this module.

satya164 replied:
@foghina Yes, we should. I'll send another PR with the changes then, after this is merged.

foghina commented:
@facebook-github-bot shipit

foghina commented:
@satya164 heads-up: patch failed internally because we don't mirror `website/`, so I had to remove the changes to that folder. Once this lands and is synced back you'll have to reapply your changes to `website/`. Sorry :(

satya164 commented:
@foghina No probs. I'll send another PR with the website changes :)
